lib/dal/src/job/definition/action.rs

```diff
@@ -121,99 +118,185 @@ impl TryFrom<JobInfo> for ActionJob {
+        .await?
+        .ok_or(ActionError::ComponentNotFoundForAction(action_id))?;
+    let component = Component::get_by_id(&ctx, component_id).await?;
+
+    if let Some(resource) = maybe_resource {
+        // Set the resource if we have a payload, regardless of status *and* assemble a
+        // summary
+        if resource.payload.is_some() {
+            component.set_resource(&ctx, resource.into()).await?;
+        }
+
+        if resource.status == ResourceStatus::Ok {
+            if resource.payload.is_none() {
+                // Clear the resource if the status is ok and we don't have a payload. This could
+                // be from invoking a delete action, rather than deleting the component.
+                component.clear_resource(&ctx).await?;
+
+                if component.to_delete() {
+                    Component::remove(&ctx, component.id()).await?;
+                }
+            }
+
+            // Remove `ActionId` from graph as the execution succeeded
+            ctx.workspace_snapshot()?
+                .remove_node_by_id(ctx.change_set()?, action_id)
+                .await?;
 
-    if matches!(status.as_ref().map(|r| r.status), Some(ResourceStatus::Ok)) {
-        let triggered_prototypes =
-            ActionPrototype::get_prototypes_to_trigger(ctx, prototype_id).await?;
-        for dependency_prototype_id in triggered_prototypes {
-            Action::new(ctx, dependency_prototype_id, Some(component_id)).await?;
+            let triggered_prototypes =
+                ActionPrototype::get_prototypes_to_trigger(&ctx, prototype.id()).await?;
+            for dependency_prototype_id in triggered_prototypes {
+                Action::new(&ctx, dependency_prototype_id, Some(component_id)).await?;
+            }
+        } else {
+            // If status is not ok, set action state to failed
+            Action::set_state(&ctx, action_id, ActionState::Failed).await?;
         }
+    } else {
+        // If the maybe_resource is none, set action state to failed
+        Action::set_state(&ctx, action_id, ActionState::Failed).await?;
     }
 
-    Ok(status)
+    let summary =
+        SummaryDiagramComponent::assemble(&ctx, &component, ChangeStatus::Unmodified).await?;
+    WsEvent::resource_refreshed(&ctx, summary)
+        .await?
+        .publish_on_commit(&ctx)
+        .await?;
+
+    WsEvent::action_list_updated(&ctx)
+        .await?
+        .publish_on_commit(&ctx)
+        .await?;
+
+    ctx.commit().await?;
+
+    Ok(())
 }
 
 #[instrument(name = "action_job.process_failed_action", skip_all, level = "info")]
-async fn process_failed_action(
-    ctx: &DalContext,
-    id: ActionId,
-    _error_message: String,
-) -> JobConsumerResult<()> {
-    info!(%id, "processing action failed");
+async fn process_failed_action(ctx: &DalContext, action_id: ActionId) -> JobConsumerResult<()> {
+    info!(%action_id, "processing action failed");
 
-    let component_id = Action::component_id(ctx, id)
```

Comment:
one thing to note - this is our last will and testament if any of the retries/commits/anything fails above. This should be as slim of an update as possible, so we will only mark the Action as failed in this case (letting the user inspect what happened and determine what to do next)